### PR TITLE
WIP : fix #1424 - issue with union and limit

### LINKED ISSUE #1424: q.take(n) ++ q produces invalid query
Minimal example for Slick 3.1.1:

``` scala
import slick.driver.PostgresDriver.api._
import slick.jdbc.JdbcBackend.Database

import scala.concurrent.Await
import scala.concurrent.duration.Duration

case class Foo(id: Int)

class Foos(tag: Tag) extends Table[Foo](tag, "foos") {
  def id = column[Int]("id", O.PrimaryKey, O.AutoInc)

  def * = id <> (Foo.apply, Foo.unapply)
}

object TestRun extends App {
  val foos = TableQuery[Foos]

  val db = Database.forURL("jdbc:postgresql://localhost/test?user=test")

  /*Await.result(
    db.run(foos.schema.create),
    Duration.Inf
  )*/

  val query = foos.take(1) ++ foos

  val result = Await.result(
    db.run(query.result),
    Duration.Inf
  )

  println(result)
}
```

This generates the query `select "id" as s17 from "foos" limit 1 union all select "id" as s17 from "foos"`, which PostgreSQL does not accept:

```
Exception in thread "main" org.postgresql.util.PSQLException: ERROR: syntax error at or near "union"
```

### PR DESCRIPTION
**Pull request by maquessime**
union with limit was not properly supported by at least 
- postgres
- derby 
- sqlite
- Hsqldb
- H2

General fix is simply to add parenthesis on both side, special case must be done for derby and sqlite

derby related issue : https://issues.apache.org/jira/browse/DERBY-2374
sqlite related issue : http://stackoverflow.com/questions/10812910/combine-two-statements-with-limits-using-union

### CODE REVIEW

File: slick-testkit/src/main/scala/com/typesafe/slick/testkit/tests/UnionTest.scala [hunk 1]
```diff
@@ -56,17 +67,26 @@ class UnionTest extends AsyncTest[RelationalTestDB] {
     } yield ()) andFinally (managers.schema ++ employees.schema).drop
   }
 
+  def testUnionWithLimit = {
```
Comment (maquessime) [inline slick-testkit/src/main/scala/com/typesafe/slick/testkit/tests/UnionTest.scala]: only test that has been added, other is just little bit of clean up

**Comment by maquessime** [comment]
wit until I fix the failing tests

File: project/Build.scala [hunk 2]
```diff
@@ -38,7 +38,7 @@ object SlickBuild extends Build {
     val testDBs = Seq(
       h2,
       "org.xerial" % "sqlite-jdbc" % "3.8.7",
-      "org.apache.derby" % "derby" % "10.9.1.0",
+      "org.apache.derby" % "derby" % "10.10.1.1",
```
Comment (maquessime) [inline project/Build.scala]: I've bump the version because I was getting trouble with an order by for derby, seems to be fixed in this version

**Comment by maquessime** [comment]
@cvogt @radsaggi do travis and circleCI run the same tests?

**Comment by cvogt** [comment]
@maquessime it seems that circle is not running the mysql and postgres tests. Was there any reasons for that @szeiger? Did it not support that when we set it up? Interestingly it does start mysql and postgres, but does not run the corresponding tests

**Comment by cvogt** [comment]
interestingly the derbydisk test also fails on Travis with a timeout but passed on circle

**Comment by maquessime** [comment]
@cvogt I'm kinda stuck here 😄 

Timeout is clearly related to the fact that I've bump derby version in tests (no timeout was defined in the previous version). There are some logged issues on derby project about login timeout.

I had to change the version because I had a trouble with an order by on one of the query (was not supported by derby 10.9.1.0) and order by was needed by sqlite otherwise, not the right result was returned.

Then, next step is the find where the login timeout of 1 sec is defined 😄

**Comment by cvogt** [comment]
good idea @maquessime . Let use know if you find anything. @radsaggi wanted to take a look as well I think.

**Comment by radsaggi** [comment]
@maquessime From what I gather, the failures are flaky. 
[This](https://issues.apache.org/jira/browse/DERBY-6250) might be a reason for the behaviour. Did you try updating derby to 10.12.1.1??

**Comment by maquessime** [comment]
@radsaggi I agree that could be the reason

With 10.12.1.1 I've got : 

```
Caused by: java.security.AccessControlException: access denied org.apache.derby.security.SystemPermission( "engine", "usederbyinternals" )
```

I've not push forward since not particular fix related to my flaky timeout issue has been done in 10.12.1.1

**Comment by cvogt** [comment]
@radsaggi @maquessime does flaky mean we sometimes can get it to succeed?

**Comment by maquessime** [comment]
@cvogt flaky enough to pass locally but fail on build machine

@radsaggi @cvogt have you run the build on this branch locally? is it fine?

Worst case, we could remove the test coverage, that really sucks but at least we will get the fix. Is it possible to discard derby only for this test? I know that's really lame :(

**Comment by radsaggi** [comment]
@cvogt @maquessime By flaky I meant the failure is unpredictable and may or may not occur. I noticed this in the rebuild I triggered for Travis. The failure occurs in different tests.

Looking at derby jira, I found [this](https://issues.apache.org/jira/browse/DERBY-6250) which might apparently be the cause(i just gave it a passing glance). If so, we might have to do with a workaround. @maquessime I could delve further into this...

**Comment by cvogt** [comment]
So for `Test com.typesafe.slick.testkit.tests.JdbcTypeTest.testBlob[derbymem]` this fails with
```
Exception in thread "AsyncExecutor.default-8" java.lang.IllegalArgumentException: requirement failed: count cannot be decreased
	at scala.Predef$.require(Predef.scala:219)
	at slick.util.ManagedArrayBlockingQueue$$anonfun$decreaseInUseCount$1.apply$mcV$sp(ManagedArrayBlockingQueue.scala:53)
	at slick.util.ManagedArrayBlockingQueue$$anonfun$decreaseInUseCount$1.apply(ManagedArrayBlockingQueue.scala:52)
	at slick.util.ManagedArrayBlockingQueue$$anonfun$decreaseInUseCount$1.apply(ManagedArrayBlockingQueue.scala:52)
	at slick.util.ManagedArrayBlockingQueue.slick$util$ManagedArrayBlockingQueue$$locked(ManagedArrayBlockingQueue.scala:217)
	at slick.util.ManagedArrayBlockingQueue.decreaseInUseCount(ManagedArrayBlockingQueue.scala:52)
	at slick.util.AsyncExecutor$$anon$2$$anon$1.afterExecute(AsyncExecutor.scala:81)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1150)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
```

as I can see.

**Comment by cvogt** [comment]
@szeiger any idea what might cause this?

@radsaggi doesn't look like the time out issue is there anymore

**Comment by radsaggi** [comment]
@maquessime It apparently works beautifully with derby 10.10.1.1. Check [this](https://github.com/radsaggi/slick/tree/issue/1424) out!
cc @cvogt

**Comment by maquessime** [comment]
@radsaggi I don't want to break your enthusiasm but I've bump to 10.11.1.1 because it has failed with 10.10.1.1 the first time : https://github.com/slick/slick/pull/1538/commits/f3ff5ac9db9ee31a52d13ffe580e18564f914805

sorry :(  random test failing is always a pain

**Pull request merged**